### PR TITLE
ceph-objectstore-tool: do not populate snapmapper with missing clones

### PR DESCRIPTION
**Pull request by liewegas**
Cache pools may have clones represented in SnapSet that are not stored
locally.  This is fine as long as ceph-objectstore-tool doesn't go and
add them when it shouldn't.

Fixes: http://tracker.ceph.com/issues/19943
Signed-off-by: Sage Weil <sage@redhat.com>

### CODE REVIEW

**Comment by liewegas** [comment]
Good catch, updated.  bluestore's douts are all one-offs, no loops.

**Comment by gregsfortytwo** [comment]
Reviewed-by: Greg Farnum <gfarnum@redhat.com> on the non-bluestore bits. (I'm just not sticking my nose in there :))

File: src/osd/SnapMapper.cc [hunk 1]
```diff
@@ -240,6 +255,11 @@ void SnapMapper::add_oid(
        ++i) {
     to_add.insert(to_raw(make_pair(*i, oid)));
   }
+  if (g_conf->subsys.should_gather(ceph_subsys_osd, 20)) {
```
Comment (gregsfortytwo) [inline src/osd/SnapMapper.cc]: Aside: I think in the past you've done this exclusion with some macro shenanigans, enclosing the whole loop in a dout/dendl pair. This format here is perfectly (perhaps more) readable, but did you switch for any particular reason?

**Comment by gregsfortytwo** [comment]
Reviewed-by: Greg Farnum <gfarnum@redhat.com>
...on the non-bluestore bits. (I'm just not sticking my nose in there :))

**Comment by liewegas** [comment]
Wrapping the whole thing with one dout/dendl works fine to generate one log entry but if there are newlines you don't get the prefix for each line.

File: src/tools/ceph_objectstore_tool.cc [hunk 2]
```diff
@@ -861,14 +861,24 @@ int get_attrs(
 	  cout << "snapset " << snapset << std::endl;
 	  if (!snapset.is_legacy()) {
 	    for (auto& p : snapset.clone_snaps) {
-	      hobject_t clone = hoid.hobj;
-	      clone.snap = p.first;
+	      ghobject_t clone = hoid;
+	      clone.hobj.snap = p.first;
 	      set<snapid_t> snaps(p.second.begin(), p.second.end());
+	      if (!store->exists(coll, clone)) {
```
Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: You could clean-up last_clones since it is not used.  last_clones is cleared when we import the head object which would be the point where it could have been useful.  It is better to check the store since you might import into a pg that already has existing objects.

Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: You can remove last_head too.  Just replace "hold == last_head" test with hold.is_head().

Comment (dzafman) [inline src/tools/ceph_objectstore_tool.cc]: 241f8ad6e2a7df739e2154ea41a1884c78502166 has the change you can cherry-pick here.

Comment (liewegas) [inline src/tools/ceph_objectstore_tool.cc]: Let's put this in a separate pr, since this one already went through testing and passed?

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-21_16:36:34-rados-wip-sage-testing-distro-basic-smithi/
